server/middleware/validationMiddleware.ts

```diff
@@ -21,6 +21,30 @@ export const findError = (errors: fieldErrors, fieldName: string) => {
   }
 }
 
+export const createSchema = <T = object>(shape: T) => zodAlwaysRefine(zObjectStrict(shape))
+
+const zObjectStrict = <T = object>(shape: T) => z.object({ _csrf: z.string().optional(), ...shape }).strict()
+
+const zodAlwaysRefine = <T extends z.ZodTypeAny>(zodType: T) =>
```

Comment:
Can you explain these changes? I'm not sure I understand them fully, apologies :)

Reply:
zObjectStrict is just a helper function to add _csrf and strict restriction because these two are applicable to all schema.

zodAlwaysRefine. that is to resolve the problem about multi-field validation. we do multi-field validation with a refine call on the root object, and the problem we encountered was that when any field in the object fails the validation, then the refine call on the root object won't be trigger, meaning we will miss that validation.
this is a helper function to work around that by doing a safeParse and adding any error to the ctx. if there is no error, the parse result will be used as transform result. if there is error, parse result will be null, and the original input (req.body) will be returned and used for subsequent refine on the root object.

Reply:
this avoid the need for working out complicated merge/union/intersection/and to build the schema.

Reply:
Can you show me a place in the code/journey at the moment where this is a problem so I can understand better? As far as I've seen we've been outputting the correct error messages where more than one field fails validation.

Reply:
`server/journeys/referral/details/schemas.ts` is where it was an issue and you have to change the schema structure in #109 

`server/journeys/referral/involvement/schemas.ts` also had challenges with it, hence you had to put them in separate object and then merge them into one schema. 

this helper function makes sure we always only need one object and don't need to merge multi objects to work around the multi field validation issue.

Reply:
Those two were to ensure the validation messages were output in the correct order - in the end it's just a simple `and` to get them in the right order, no?

Reply:
after an `and` you can't `strict()` it. also stuff like merge/and seems to mess up schema type inference.

Reply:
it wasn't just ordering.
before your fix on `server/journeys/referral/details/schemas.ts`, if involvementType input fails the validation (no valid option selected from dropdown), then the validation on hour+minute fields will not be executed.